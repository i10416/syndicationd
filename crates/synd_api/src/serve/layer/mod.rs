pub mod authenticate;
pub mod trace;
