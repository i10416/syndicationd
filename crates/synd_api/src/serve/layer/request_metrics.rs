use std::{
    convert::Infallible,
    pin::Pin,
    task::{Context, Poll},
    time::Instant,
};

use axum::{extract::Request, http::Method, response::Response};
use futures_util::Future;
use synd_o11y::metric;
use tower::{Layer, Service};

#[derive(Clone)]
pub struct RequestMetricsLayer {}

impl RequestMetricsLayer {
    pub fn new() -> Self {
        Self {}
    }
}

impl<S> Layer<S> for RequestMetricsLayer {
    type Service = RequestMetricsService<S>;

    fn layer(&self, inner: S) -> Self::Service {
        RequestMetricsService { inner }
    }
}

#[derive(Clone)]
pub struct RequestMetricsService<S> {
    inner: S,
}

impl<S> Service<Request> for RequestMetricsService<S>
where
    S: Service<Request, Response = Response, Error = Infallible> + Clone + Send + 'static,
    S::Future: Send,
{
    type Response = Response;
    type Error = Infallible;
    type Future =
        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request) -> Self::Future {
        let start = Instant::now();
        let path = req.uri().path().to_owned();
        let method = req.method().clone();

        let mut this = self.clone();
        Box::pin(async move {
            let response = this.inner.call(req).await.unwrap();
            let status = response.status().as_u16();

            // https://opentelemetry.io/docs/specs/semconv/http/http-metrics/
            // Considiering the case of not found(404), recording the path as
            // an attribute leads to an inability to control cardinality.
            // Therefore, the path is not recorded.
            metric!(
                monotonic_counter.http.server.request = 1,
                http.response.status.code = status
            );

            // instrument graphql latency
            if path == "/graphql" && method == Method::POST {
                // f64 is matter
                // The type of the field that MetricsVisitor visits when on_event() is called is pre defined.
                // If u128 which is returned from elapsed() is used, it will not be visited, resulting in no metrics recorded.
                // Spec say "When instruments are measuring durations, seconds SHOULD be used"
                // https://opentelemetry.io/docs/specs/semconv/general/metrics/#instrument-units
                let elapsed: f64 = start.elapsed().as_secs_f64();
                // is there any semantic conventions?
                metric!(histogram.graphql.duration = elapsed);
            }

            Ok(response)
        })
    }
}
