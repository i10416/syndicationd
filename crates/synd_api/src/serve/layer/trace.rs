use tower_http::trace::HttpMakeClassifier;
use tracing::Level;

#[derive(Clone)]
pub struct MakeSpan;

impl<B> tower_http::trace::MakeSpan<B> for MakeSpan {
    fn make_span(&mut self, request: &axum::http::Request<B>) -> tracing::Span {
        use synd_o11y::opentelemetry::extension::*;
        let cx = synd_o11y::opentelemetry::http::extract(request.headers());

        let request_id = cx
            .baggage()
            .get("request.id")
            .map_or("?".into(), |v| v.as_str());

        let span = tracing::span!(
            Level::INFO,
            "http",
            method = %request.method(),
            uri = %request.uri(),
            %request_id,
        );

        tracing::info!("Request headers: {:#?}", request.headers());

        span.set_parent(cx);
        span
    }
}

#[derive(Clone)]
pub struct OnRequest;

impl<B> tower_http::trace::OnRequest<B> for OnRequest {
    fn on_request(&mut self, _request: &axum::http::Request<B>, _span: &tracing::Span) {
        // do nothing
    }
}

pub fn layer() -> tower_http::trace::TraceLayer<
    HttpMakeClassifier,
    MakeSpan,
    OnRequest,
    tower_http::trace::DefaultOnResponse,
> {
    tower_http::trace::TraceLayer::new_for_http()
        .make_span_with(MakeSpan)
        .on_request(OnRequest)
        .on_response(tower_http::trace::DefaultOnResponse::default().level(Level::INFO))
}
