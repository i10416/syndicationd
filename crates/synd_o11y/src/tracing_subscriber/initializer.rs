use crate::{
    opentelemetry::init_propagation,
    opentelemetry_layer,
    tracing_subscriber::{audit, otel_metrics::metrics_event_filter},
    OpenTelemetryGuard,
};

const LOG_DIRECTIVE: &str = "SYND_LOG";
const DEFAULT_LOG_DIRECTIVE: &str = "info";

pub struct TracingInitializer {
    app_name: Option<&'static str>,
    app_version: Option<&'static str>,
    otlp_endpoint: Option<String>,
    trace_sampler_ratio: f64,
    enable_ansi: bool,
    show_code_location: bool,
    show_target: bool,
}

impl Default for TracingInitializer {
    fn default() -> Self {
        Self {
            app_name: None,
            app_version: None,
            otlp_endpoint: None,
            trace_sampler_ratio: 1.,
            enable_ansi: true,
            show_code_location: false,
            show_target: true,
        }
    }
}

impl TracingInitializer {
    /// Initialize tracing Subscriber with layers
    pub fn init(self) -> Option<OpenTelemetryGuard> {
        use tracing_subscriber::{
            filter::EnvFilter, fmt, layer::SubscriberExt, util::SubscriberInitExt as _, Layer as _,
            Registry,
        };

        let Self {
            app_name: Some(app_name),
            app_version: Some(app_version),
            otlp_endpoint,
            trace_sampler_ratio,
            enable_ansi,
            show_code_location,
            show_target,
        } = self
        else {
            panic!()
        };

        let (otel_layer, otel_guard) =
            match otlp_endpoint
                .as_deref()
                .filter(|s| !s.is_empty())
                .map(|endpoint| {
                    opentelemetry_layer(endpoint, app_name, app_version, trace_sampler_ratio)
                }) {
                Some((otel_layer, otel_guard)) => (Some(otel_layer), Some(otel_guard)),
                _ => (None, None),
            };

        Registry::default()
            .with(
                fmt::Layer::new()
                    .with_ansi(enable_ansi)
                    .with_timer(fmt::time::ChronoLocal::rfc_3339())
                    .with_file(show_code_location)
                    .with_line_number(show_code_location)
                    .with_target(show_target)
                    .with_filter(metrics_event_filter())
                    .and_then(otel_layer)
                    .with_filter(
                        EnvFilter::try_from_env(LOG_DIRECTIVE)
                            .or_else(|_| EnvFilter::try_new(DEFAULT_LOG_DIRECTIVE))
                            .unwrap()
                            .add_directive(audit::Audit::directive()),
                    ),
            )
            .with(audit::layer())
            .init();

        // Set text map propagator globally
        init_propagation();

        otel_guard
    }
}

impl TracingInitializer {
    #[must_use]
    pub fn app_name(self, app_name: &'static str) -> Self {
        Self {
            app_name: Some(app_name),
            ..self
        }
    }

    #[must_use]
    pub fn app_version(self, app_version: &'static str) -> Self {
        Self {
            app_version: Some(app_version),
            ..self
        }
    }

    #[must_use]
    pub fn otlp_endpoint(self, otlp_endpoint: Option<String>) -> Self {
        Self {
            otlp_endpoint,
            ..self
        }
    }

    #[must_use]
    pub fn trace_sampler_ratio(self, trace_sampler_ratio: f64) -> Self {
        Self {
            trace_sampler_ratio,
            ..self
        }
    }

    #[must_use]
    pub fn enable_ansi(self, enable_ansi: bool) -> Self {
        Self {
            enable_ansi,
            ..self
        }
    }

    #[must_use]
    pub fn show_code_location(self, show_code_location: bool) -> Self {
        Self {
            show_code_location,
            ..self
        }
    }

    #[must_use]
    pub fn show_target(self, show_target: bool) -> Self {
        Self {
            show_target,
            ..self
        }
    }
}
