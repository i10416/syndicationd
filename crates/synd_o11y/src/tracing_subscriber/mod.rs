use std::borrow::Cow;

use tracing::Subscriber;
use tracing_subscriber::{registry::LookupSpan, Layer};

pub mod audit;
pub mod otel_log;
pub mod otel_metrics;
pub mod otel_trace;

/// Return a composed layer that exports traces, metrics, and logs of opentelemetry.
/// The exported telemetry will be accompanied by a `Resource`.
pub fn opentelemetry_layer<S>(
    endpoint: impl Into<String>,
    service_name: impl Into<Cow<'static, str>>,
    service_version: impl Into<Cow<'static, str>>,
    trace_sampler_ratio: f64,
) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    let endpoint = endpoint.into();
    let resource = crate::opentelemetry::resource(service_name, service_version);

    otel_trace::layer(endpoint.clone(), resource.clone(), trace_sampler_ratio)
        .and_then(otel_metrics::layer(endpoint.clone(), resource.clone()))
        .and_then(otel_log::layer(endpoint, resource))
}
