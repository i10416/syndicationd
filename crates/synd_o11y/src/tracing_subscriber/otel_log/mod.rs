use opentelemetry_appender_tracing::layer::OpenTelemetryTracingBridge;
use opentelemetry_otlp::WithExportConfig as _;
use opentelemetry_sdk::{runtime, Resource};
use tracing::Subscriber;
use tracing_subscriber::{registry::LookupSpan, Layer};

pub fn layer<S>(endpoint: impl Into<String>, resource: Resource) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    opentelemetry_otlp::new_pipeline()
        .logging()
        .with_log_config(opentelemetry_sdk::logs::Config::default().with_resource(resource))
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint(endpoint),
        )
        .install_batch(runtime::Tokio)
        .unwrap();

    let provider = opentelemetry::global::logger_provider();

    OpenTelemetryTracingBridge::new(&provider)
}
