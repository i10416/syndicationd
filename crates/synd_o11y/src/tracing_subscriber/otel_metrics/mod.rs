use std::time::Duration;

use opentelemetry_sdk::{metrics::MeterProvider, runtime, Resource};
use tracing::{Metadata, Subscriber};
use tracing_opentelemetry::MetricsLayer;
use tracing_subscriber::{filter::filter_fn, layer::Filter, registry::LookupSpan, Layer};

pub mod macros;

pub const METRICS_EVENT_TARGET: &str = "metrics";

pub fn metrics_event_filter<S: Subscriber>() -> impl Filter<S> {
    filter_fn(|metadata: &Metadata<'_>| metadata.target() != METRICS_EVENT_TARGET)
}

pub fn layer<S>(resource: Resource) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    MetricsLayer::new(init_meter_provider(resource))
}

fn init_meter_provider(resource: Resource) -> MeterProvider {
    let provider = opentelemetry_otlp::new_pipeline()
        .metrics(runtime::Tokio)
        .with_resource(resource)
        .with_exporter(opentelemetry_otlp::new_exporter().tonic())
        .with_period(Duration::from_secs(60))
        .build()
        .unwrap();

    opentelemetry::global::set_meter_provider(provider.clone());

    provider
}
