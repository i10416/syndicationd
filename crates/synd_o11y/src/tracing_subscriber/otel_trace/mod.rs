use opentelemetry::{global, trace::TracerProvider};
use opentelemetry_otlp::WithExportConfig;
use opentelemetry_sdk::{
    runtime,
    trace::{BatchConfig, Sampler, Tracer},
    Resource,
};
use tracing::Subscriber;
use tracing_opentelemetry::OpenTelemetryLayer;
use tracing_subscriber::{registry::LookupSpan, Layer};

pub fn layer<S>(
    endpoint: impl Into<String>,
    resource: Resource,
    sampler_ratio: f64,
    batch_config: BatchConfig,
) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    OpenTelemetryLayer::new(init_tracer(endpoint, resource, sampler_ratio, batch_config))
}

#[expect(clippy::needless_pass_by_value)]
fn init_tracer(
    endpoint: impl Into<String>,
    resource: Resource,
    sampler_ratio: f64,
    // TODO: how to use BatchConfig after 0.27 ?
    _batch_config: BatchConfig,
) -> Tracer {
    let exporter = opentelemetry_otlp::SpanExporter::builder()
        .with_tonic()
        .with_endpoint(endpoint)
        .build()
        .unwrap();

    let provider = opentelemetry_sdk::trace::TracerProvider::builder()
        .with_resource(resource)
        .with_sampler(Sampler::ParentBased(Box::new(Sampler::TraceIdRatioBased(
            sampler_ratio,
        ))))
        .with_batch_exporter(exporter, runtime::Tokio)
        .build();

    // > It would now be the responsibility of users to set it by calling global::set_tracer_provider(tracer_provider.clone());
    //  https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-otlp/CHANGELOG.md#v0170
    global::set_tracer_provider(provider.clone());

    provider.tracer("tracing-opentelemetry")
}

#[cfg(test)]
mod tests {
    use std::{net::SocketAddr, time::Duration};

    use opentelemetry::KeyValue;
    use opentelemetry_proto::tonic::{
        collector::trace::v1::{
            trace_service_server::{TraceService, TraceServiceServer},
            ExportTraceServiceRequest, ExportTraceServiceResponse,
        },
        trace::v1::{span::SpanKind, status::StatusCode, Status},
    };
    use opentelemetry_sdk::trace::BatchConfigBuilder;
    use tokio::sync::mpsc::UnboundedSender;
    use tonic::transport::Server;
    use tracing::dispatcher;
    use tracing_subscriber::{layer::SubscriberExt, Registry};

    use super::*;

    type Request = tonic::Request<ExportTraceServiceRequest>;

    struct DumpTraces {
        tx: UnboundedSender<Request>,
    }

    #[tonic::async_trait]
    impl TraceService for DumpTraces {
        async fn export(
            &self,
            request: tonic::Request<ExportTraceServiceRequest>,
        ) -> Result<tonic::Response<ExportTraceServiceResponse>, tonic::Status> {
            self.tx.send(request).unwrap();

            Ok(tonic::Response::new(ExportTraceServiceResponse {
                partial_success: None, // means success
            }))
        }
    }

    #[tracing::instrument(fields(
        otel.name = "f1_custom",
        otel.kind = "Client",
    ) )]
    fn f1() {
        f2();
    }
    #[tracing::instrument(fields(
        otel.name = "f2_custom",
        otel.kind = "Server",
    ))]
    fn f2() {
        tracing::error!("error_f2");
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn layer_test() {
        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
        let dump = TraceServiceServer::new(DumpTraces { tx });
        let addr: SocketAddr = ([127, 0, 0, 1], 48100).into();
        let server = Server::builder().add_service(dump).serve(addr);
        let _server = tokio::task::spawn(server);
        let resource = resource();
        let config = BatchConfigBuilder::default()
            // The default interval is 5 seconds, which slows down the test
            .with_scheduled_delay(Duration::from_millis(10))
            .build();
        let layer = layer("https://localhost:48100", resource.clone(), 1.0, config);
        let subscriber = Registry::default().with(layer);
        let dispatcher = tracing::Dispatch::new(subscriber);

        dispatcher::with_default(&dispatcher, || {
            f1();
        });

        let req = rx.recv().await.unwrap().into_inner();
        assert_eq!(req.resource_spans.len(), 1);

        let resource_span = req.resource_spans[0].clone();
        insta::with_settings!({
            description => "trace resource",
        }, {
            insta::assert_yaml_snapshot!("layer_test_trace_resource", req.resource_spans[0].resource);
        });

        let [f2_span, f1_span] = resource_span.scope_spans[0].spans.as_slice() else {
            panic!()
        };

        assert_eq!(&f2_span.name, "f2_custom");
        assert_eq!(&f1_span.name, "f1_custom");
        assert_eq!(f2_span.parent_span_id, f1_span.span_id);
        assert_eq!(f2_span.trace_id, f1_span.trace_id);
        assert_eq!(f2_span.kind, SpanKind::Server as i32);
        assert_eq!(f1_span.kind, SpanKind::Client as i32);
        assert_eq!(
            f2_span.status,
            Some(Status {
                message: String::new(),
                code: StatusCode::Error as i32,
            })
        );
        assert_eq!(
            f1_span.status,
            Some(Status {
                message: String::new(),
                code: StatusCode::Unset as i32,
            })
        );
        assert_eq!(f2_span.events.len(), 1);
        assert_eq!(f2_span.attributes.len(), 7);
    }

    fn resource() -> Resource {
        Resource::new([KeyValue::new("service.name", "test")])
    }
}
