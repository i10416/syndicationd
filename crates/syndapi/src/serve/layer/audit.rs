use tracing::{
    field,
    span::{self, Attributes},
    subscriber::Interest,
    Event, Level, Metadata, Subscriber,
};
use tracing_subscriber::{
    filter::{Directive, Filtered},
    layer::{self, Context},
    registry::LookupSpan,
    Layer,
};

mod macros {

    #[macro_export]
    macro_rules! audit_span {
        () => {
            ::tracing::info_span!(
                target: $crate::serve::layer::audit::Audit::TARGET,
                $crate::serve::layer::audit::Audit::SPAN_ROOT_NAME)
        };
    }

    #[macro_export]
    macro_rules! audit {
        ($($arg:tt)*) => {
            ::tracing::event!(
                name: $crate::serve::layer::audit::Audit::EVENT_NAME,
                target: $crate::serve::layer::audit::Audit::TARGET,
                ::tracing::Level::TRACE, $($arg)*)
        };
    }
}

pub struct Audit;

impl Audit {
    pub const TARGET: &'static str = "__audit";
    pub const SPAN_ROOT_NAME: &'static str = "audit.root";
    pub const EVENT_NAME: &'static str = "audit.event";

    const EMIT_TARGET: &'static str = "audit";
    const EMIT_EVENT_NAME: &'static str = "audit";

    // follow opentelemetry semantic conventions
    pub const USER_ID: &'static str = "enduser.id";
    pub const OPERATION: &'static str = "operation";
    pub const RESULT: &'static str = "result";

    pub fn directive() -> Directive {
        let directive = format!("{emit}=info", emit = Self::EMIT_TARGET);

        directive.parse().expect("Invalid directive")
    }
}

/// Create AuditLayer
pub fn layer<S>() -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    let layer = AuditLayer::new();
    let filter = AuditFilter::new();
    Filtered::new(layer, filter)
}

pub struct AuditFilter;

impl AuditFilter {
    fn new() -> Self {
        Self {}
    }

    fn is_enabled(&self, meta: &Metadata<'_>) -> bool {
        meta.target() == Audit::TARGET
    }
}

impl<S> layer::Filter<S> for AuditFilter {
    fn enabled(&self, meta: &Metadata<'_>, _cx: &Context<'_, S>) -> bool {
        self.is_enabled(meta)
    }

    fn callsite_enabled(&self, meta: &'static Metadata<'static>) -> Interest {
        if self.is_enabled(meta) {
            Interest::always()
        } else {
            Interest::never()
        }
    }
}

pub struct AuditLayer {}

impl AuditLayer {
    fn new() -> Self {
        Self {}
    }
}

struct AuditEventVisitor<'a> {
    ctx: &'a mut AuditContext,
}

impl<'a> field::Visit for AuditEventVisitor<'a> {
    fn record_debug(&mut self, _field: &field::Field, _value: &dyn std::fmt::Debug) {
        // do nothing
    }

    fn record_str(&mut self, field: &field::Field, value: &str) {
        match field.name() {
            Audit::USER_ID => self.ctx.user_id = Some(value.to_owned()),
            Audit::OPERATION => self.ctx.operation = Some(value.to_string()),
            Audit::RESULT => self.ctx.result = Some(value.to_string()),
            _ => {}
        }
    }
}

struct AuditContext {
    user_id: Option<String>,
    operation: Option<String>,
    result: Option<String>,
}

impl AuditContext {
    fn new() -> Self {
        Self {
            user_id: None,
            operation: None,
            result: None,
        }
    }
}

impl<S> Layer<S> for AuditLayer
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    /// If new span is audit root span, create AuditContext and insert to extensions.
    fn on_new_span(&self, attrs: &Attributes<'_>, id: &span::Id, ctx: Context<'_, S>) {
        if attrs.metadata().name() != Audit::SPAN_ROOT_NAME {
            return;
        }
        let span = ctx.span(id).expect("Span not found, this is a bug");
        let mut extensions = span.extensions_mut();
        extensions.insert(AuditContext::new());
    }

    /// If event is in audit span, visit event and store audit information to extensions
    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
        if event.metadata().name() != Audit::EVENT_NAME {
            return;
        }

        // traverse span tree to find audit context
        let Some(span) = ctx.lookup_current() else {
            return;
        };
        let Some(audit_span) = span
            .scope()
            .from_root()
            .find(|span| span.metadata().name() == Audit::SPAN_ROOT_NAME)
        else {
            return;
        };
        let mut extension = audit_span.extensions_mut();
        let Some(audit_ctx) = extension.get_mut::<AuditContext>() else {
            return;
        };

        event.record(&mut AuditEventVisitor { ctx: audit_ctx });
    }

    /// If audit root span is closed, write a audit log
    fn on_close(&self, id: span::Id, ctx: Context<'_, S>) {
        let span = ctx.span(&id).expect("Span not found, this is a bug");
        if span.metadata().name() != Audit::SPAN_ROOT_NAME {
            return;
        }
        let mut extensions = span.extensions_mut();
        let Some(AuditContext {
            user_id,
            operation,
            result,
        }) = extensions.remove::<AuditContext>()
        else {
            return;
        };

        let user_id = user_id.as_deref().unwrap_or("?");
        let operation = operation.as_deref().unwrap_or("?");
        let result = result.as_deref().unwrap_or("?");

        tracing::event!(
            name: Audit::EMIT_EVENT_NAME,
            target: Audit::EMIT_TARGET,
            Level::INFO,
            { Audit::USER_ID } = user_id,
            { Audit::OPERATION } = operation,
            { Audit::RESULT } = result,
        );
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, Mutex};

    use tracing::{info, info_span};
    use tracing_subscriber::prelude::__tracing_subscriber_SubscriberExt;

    use super::*;

    struct TestLayer<F> {
        on_event: F,
    }

    impl<S, F> Layer<S> for TestLayer<F>
    where
        S: Subscriber + for<'span> LookupSpan<'span>,
        F: Fn(&Event<'_>) + 'static,
    {
        fn on_event(&self, event: &Event<'_>, _ctx: Context<'_, S>) {
            (self.on_event)(event);
        }
    }

    // Mock audit layer usecase
    mod usecase_authorize_scenario {
        use tracing::info_span;

        use crate::{audit, audit_span, serve::layer::audit::Audit};

        pub fn root() {
            let span = audit_span!();
            let _enter = span.enter();
            usecase();
        }

        fn usecase() {
            let span = info_span!("usecase");
            let _enter = span.enter();
            authorize();
            ops();
            audit!(
                { Audit::OPERATION } = "create_foo",
                { Audit::RESULT } = "success",
            );
        }

        fn authorize() {
            let span = info_span!("authorize");
            let _enter = span.enter();
            audit!({ Audit::USER_ID } = "user-a",);
        }

        fn ops() {
            let span = info_span!("ops");
            let _enter = span.enter();
        }
    }

    #[test]
    fn usecase_authorize_scenario() {
        let ctx = Arc::new(Mutex::new(AuditContext::new()));
        let ctx2 = Arc::clone(&ctx);
        let on_event = move |event: &Event<'_>| {
            if event.metadata().name() == Audit::EMIT_EVENT_NAME {
                let mut ctx = ctx2.lock().unwrap();
                event.record(&mut AuditEventVisitor { ctx: &mut ctx });
            }
        };
        let test_layer = TestLayer { on_event };
        let subscriber = tracing_subscriber::registry()
            .with(layer())
            .with(test_layer);

        tracing::subscriber::with_default(subscriber, || {
            usecase_authorize_scenario::root();
        });

        let ctx = Arc::into_inner(ctx).unwrap().into_inner().unwrap();

        assert_eq!(ctx.user_id.as_deref(), Some("user-a"));
        assert_eq!(ctx.operation.as_deref(), Some("create_foo"));
        assert_eq!(ctx.result.as_deref(), Some("success"));
    }

    #[test]
    fn no_audit_span() {
        let on_event = |event: &Event<'_>| {
            if event.metadata().name() == Audit::EMIT_EVENT_NAME {
                panic!("should not called");
            }
        };
        let test_layer = TestLayer { on_event };
        let subscriber = tracing_subscriber::registry()
            .with(layer())
            .with(test_layer);

        tracing::subscriber::with_default(subscriber, || {
            let span = info_span!("ignore");
            let _enter = span.enter();
            info!("ignore");
        });
    }

    #[test]
    fn emit_target_dose_not_equal_target() {
        assert!(Audit::TARGET != Audit::EMIT_TARGET);
    }
}
