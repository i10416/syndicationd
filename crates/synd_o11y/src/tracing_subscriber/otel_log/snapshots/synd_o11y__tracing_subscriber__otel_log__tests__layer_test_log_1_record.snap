---
source: crates/synd_o11y/src/tracing_subscriber/otel_log/mod.rs
description: " log 1 record"
expression: record
---
timeUnixNano: "0"
observedTimeUnixNano: "[OBSERVED_TIME_UNIX_NANO]"
severityNumber: 9
severityText: INFO
body:
  stringValue: f1_message
attributes: []
droppedAttributesCount: 0
flags: 0
traceId: ""
spanId: ""
